### ISSUE TITLE
Defunct processes increasing on very busy server.

### ISSUE THREAD

**Issue by laapsaap** [grke/burp#316]
Hi,

I have a serious problem with burp right now. The server where I run burp server on has a very high "fictional" load.  10:42:13 up 17:33,  2 users,  load average: 118.10, 117.99, 116.22.   But in reality CPU usage is <1%, mem is 50% and disk io is 4kb/sec.

```
[al:root@135 ~] (backup) # ps aux | grep burp | wc -l
114
[al:root@135 ~] (backup) # killall -9 burp
[all:root@135 ~] (backup) # ps aux | grep burp | wc -l
114
```

```
root     20089  0.0  0.0  67272  7252 ?        D    09:55   0:00 burp -c /etc/burp/burp-server.conf
root     21119  0.3  0.0  63748  4012 ?        D    Aug06   3:30 burp -c /etc/burp/burp-server.conf
root     21342  0.0  0.0  67272  7252 ?        D    10:05   0:00 burp -c /etc/burp/burp-server.conf
```
Thats because the burp processes all run as D state (uninterruptable sleep).

I have no idea what is causing this, but the only remedy is to power cycle the server.

**Comment by tokudan** [comment]
Preventing a SIGKILL is only possible in the kernel, no process can catch that. So you're facing an issue that's not in burp.
From my experience with similar issues I'd assume you have burp accessing some kind of NFS server and haven't taken care of the different mount options (soft/hard/intr), but it also could be another filesystem or another kernel function blocking the kill.
As soon as you resolve the underlying issue, the burp processes will be killed by the kernel.

**Comment by grke** [comment]
I always thought 'D' meant 'disk wait'. I see it on processes waiting on disk access. Especially on my slow USB disk, or when there is some network mount problem.

**Comment by laapsaap** [comment]
We use local disks. I get defunct burp procceses after awhile, but I always ignored it, not sure if this is related.

PROCESS STATE CODES
       Here are the different values that the s, stat and state output specifiers (header "STAT" or "S") will display to describe the state of
       a process.
       D    Uninterruptible sleep (usually IO)

I understand I cannot kill burp because its waiting on something in kernel space, but shouldnt burp timeout at some point?

**Comment by tokudan** [comment]
burp cannot timeout. The kernel function would have to timeout. If the processes are stuck like this for a long time it seems to be a function that does not timeout.
I wouldn't worry about the zombie processes, they have exited already, just their parent process has not cleaned them up, so the kernel cannot release the structures in the process tables yet. With the parent process hung in the kernel somewhere, that's pretty much expected.

Since it seems to be reproducible, you could try running burp with strace, to trace all system calls. When you notice burp hanging again, you can just check the output from strace and see which system call hangs.
something like this should do the trick:
```
# strace -o /tmp/burp.strace -f -ff /usr/sbin/burp -c /etc/burp/burp-server.conf
```
Once you notice burp hanging, just check the log file for that pid (/tmp/burp.strace.PID). You might want to choose a directory with some space available, strace log files can get rather big if there's lots of data.

**Comment by grke** [comment]
How many clients do you have trying to connect at the same time?
Are you using protocol 1 or 2?

**Comment by laapsaap** [comment]
I am using protocol 1, we have 300 clients connecting in a 5 minute time frame :)   Should increase the time frame. I do average 15 backups simultaneously.

**Comment by grke** [comment]
I think you definitely have too many clients connecting simultaneously. You should spread them out some more and maybe consider the 'randomise' client option.
Especially if they are currently all connecting at the same moment.

Even if the burp server is rejecting most connections, it still has to process each one in order to reject it, so the fewer at the same time, the better.

**Comment by laapsaap** [comment]
yeah that is a smart thing to do, will increase connection interval. But the defunct processes are a bit annoying, I know it doesnt affect performance. But is there anything I can do to prevent his?

This is one week running burp service
```
[al:root@backupserver ~] (backup) # ps aux | grep defunct | wc -l
1951
```

**Comment by grke** [comment]
On Tue, August 18, 2015 9:36 pm, laapsaap wrote:
> yeah that is a smart thing to do, will increase connection interval. But
> the defunct processes are a bit annoying, I know it doesnt affect
> performance. But is there anything I can do to prevent his?
>
> This is one week running burp service
> ```
> [al:root@backupserver ~] (backup) # ps aux | grep defunct | wc -l
> 1951
> ```
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/grke/burp/issues/316#issuecomment-132182274

Hello,

I have a patch for the server that might help.
I think that multiple SIGCHLDs on a busy server are getting conflated into
one, meaning that only one waitpid() gets run on multiple children exiting
at the same time, and therefore the defunct list grows.

I have attached a patch to this message, as I am unable to commit to github
until I get internet access at home.

**Comment by grke** [comment]
Hmm, I can't see the attachment in github. I will email you directly.

**Comment by grke** [comment]
Actually, I have put the patch on my website. You can get it here:

http://grke.net/waitpid.patch

**Comment by grke** [comment]
Minor issue in the first patch. Try this one instead:

http://grke.net/waitpid2.patch

**Comment by grke** [comment]
Patch merged into git master.

**Comment by laapsaap** [comment]
Thank you for all the fixes, didnt have time to test your fixes yet. But under the circumstances, I am satisfied with running burp 2.0.21 in production. Doing 200 clients a day (linux and macs). Will compile master this week and test it. Thanks

**Comment by laapsaap** [comment]
Still happening with 2.0.22 release. Updated all the clients and server instance.

0:00 [burp] <defunct>

**Comment by grke** [comment]
You will still have defunct ones until the server cleans them up, that is normal.
The question is whether you get more of them hanging around as time goes on.

**Comment by laapsaap** [comment]
Sorry, I forgot to mention I rebooted the server. These are new zombie processes.

**Comment by grke** [comment]
I am now able to reproduce this in two different ways, so I'll try to fix it again when I have a spare moment.

**Comment by grke** [comment]
Another fix is now in master, which solves the two different ways that I could reproduce the problem.

**Comment by laapsaap** [comment]
Looks good! Burp is running for over an hour now and I dont see any defunct processes. Normally I will have 40-50 already. Thanks
